torch_xla/distributed/xla_dist.py

```diff
@@ -656,14 +654,15 @@ def _gcloud_scp(local_path, remote_path, client_worker):
   def _cleanup(self, script_map):
 
     def _cleanup_worker(local_script, remote_script, client_worker):
-      rm_script = ['rm', remote_script]
-      self._build_and_run_ssh(rm_script, client_worker)
-      subprocess.call(['rm', local_script])
+      rm_tmp_dir = ['rm', '-rf', os.path.dirname(remote_script)]
+      self._build_and_run_ssh(rm_tmp_dir, client_worker)
+      subprocess.call(['rm', '-rf', os.path.dirname(local_script)])
       if self.docker_image:
         rm_container = ['docker', 'rm', '-f', self.docker_container]
         self._build_and_run_ssh(rm_container, client_worker)
-      rm_proc = ['pkill', '-u', self.DEFAULT_USER_NAME]
-      self._build_and_run_ssh(rm_proc, client_worker, log=False)
+      rm_pgroup = ('kill -9 -$(ps xao pid,pgid,cmd | grep {} | grep -v grep'
```

Comment:
This looks a bit `Meh` to me, so let's think for a bit more aiming at a better solution.

Reply:
Maybe using `psutil` instead of that shell mambo-jumbo? 😄

Reply:
I tried a couple of things (bug 145665509). Do you happen to have any other ideas?

Reply:
Ah just saw your comment on `psutil`. I wasn't aware of it so thanks for the suggestion! I'll check it out on Monday.

Notes for self:
* Looks like `psutil.process_iter()` ([API](https://psutil.readthedocs.io/en/latest/#psutil.process_iter)) would do what I want since `Process.gids()` seems to get the `$PGID`.
* Also `Process.cmdline()` should contain the remote script name so I'll use that to select correct process' pgroup to kill.
* Then `os.killpg($PGID, signal.SIGTERM)` should do the job, but to verify whether it also kills children, grandchildren, etc.
* All this functionality will then be another module in somewhere like `torch_xla.utils.pkill` or stuff this in the remote command sent with something like `python -c 'import psutil;....'`

Reply:
Oh, duh!
Never mind, I thought this was just python being generated.
I wonder if we should have instead a python script template that we build, which will allow us to do a bit more complex things.
In there we can use all the facilities we want, like psutil, etc...
This will make things easier IMHO, when the stuff we have to do there becomes a bit less trivial than running commands.

Given the template. we could use string substitution, or even simpler, just prepend the required "arguments".
Like:
```Python
# Prepended arguments
CMDLINE = [...]
SOME_ARG = ...
...

# Python script template starts here
import psutil
import subprocess

...
```

Reply:
I think what you describe is mentioned in the above [comment](https://github.com/pytorch/xla/pull/1469#discussion_r355143659) so I think we're on the same page 😄. I'll do add the killing module under `torch_xla.utils.pkill` and issue that remote cleanup command instead.

Reply:
Updated to use `psutil` and tested after building `torch_xla` wheel and installing it on 4 VMs with v*-32 slice.

Reply:
> I think what you describe is mentioned in the above [comment](https://github.com/pytorch/xla/pull/1469#discussion_r355143659) so I think we're on the same page 😄. I'll do add the killing module under `torch_xla.utils.pkill` and issue that remote cleanup command instead.

That is not what I meant 😉 
I meant, instead of having a shell script being generated, have a python script generated from a template.
You read the template, and create the real script by appending the template to a set of Python variables the template will use (see above in my example).
The template can then use subprocess.call() and any other python facility to do stuff, as well as psutil killing.

Reply:
Per offline conversation marking, reverted to be all bash-based.